### ISSUE TITLE
fused-Adam path silently zeroes NaN loss readout when trace lambda has many params + CCE-style chain

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#396]
## Summary

Reported by AiDotNet#1346 follow-up investigation. `CompiledTrainingPlan<T>.Step()` returns `LastLoss = 0` (literal float zero, not NaN, not Inf) when:
- The traced forward+loss lambda routes through a CCE-style chain (`TensorAddScalar` → `TensorLog` → `TensorMultiply` → `ReduceSum` → `ReduceMean` → `TensorNegate`)
- The predicted input to CCE contains negative values (so `log(p + eps)` produces NaN inside the chain)
- The plan is configured for fused Adam updates via `ConfigureOptimizer(OptimizerType.Adam, ...)`
- The model has many trainable parameters (Tensors-level minimal traces with just `[W, b]` do NOT reproduce — see "Bisection" below)

The consumer sees `GetLastLoss() == 0` and assumes training is converging while gradients are corrupted by NaN. The model output stays at random init forever. This is the "silent degradation" symptom behind AiDotNet#1346 (FlashAttention PathB sanity test stayed at top1=0% / top5=100% / ppl=V after engine fix #362 landed).

The bug surface is the **AiDotNet→Tensors loss-readout path** under specific compile-time conditions. Tensors-side direct equivalents (same op chain, fewer params, no AiDotNet) correctly propagate the value (NaN or otherwise) all the way through to `lossOutput[0]`.

## Reproducer

The minimal repro is the AiDotNet integration test below (file: `tests/AiDotNet.Tests/IntegrationTests/NeuralNetworks/FlashAttentionFusedCompiledTrainingIssue1346Tests.cs` from the diagnostic branch `fix/1346-flashattention-compile-backward`). It fails on **AiDotNet master @ 68f5c69 + AiDotNet.Tensors NuGet 0.81.3** (which includes engine-fix PR #362 + the 5-op audit PR #367):

```csharp
[Fact]
public async Task DenseIdentity_CCE_OnFusedAdam_NaNOrZero_PinsLastLossBug()
{
    AiDotNet.Training.CompiledTapeTrainingStep<float>.ResetFusedStepCount();
    AiDotNet.Training.CompiledTapeTrainingStep<float>.Invalidate();

    var layers = new List<ILayer<float>>
    {
        new DenseLayer<float>(16, new IdentityActivation<float>()),
        new LayerNormalizationLayer<float>(),
        new SequenceTokenSliceLayer<float>(SequenceTokenSliceLayer<float>.Position.Last),
        new DenseLayer<float>(8, new IdentityActivation<float>())  // raw logits
    };
    var arch = new TransformerArchitecture<float>(/* ... 4 seq * 16 embed * 8 classes ... */, layers: layers);
    var model = new Transformer<float>(arch,
        lossFunction: new CategoricalCrossEntropyLoss<float>(),
        optimizer: new AdamOptimizer<float, Tensor<float>, Tensor<float>>(null, /* lr=0.01 */ ...));
    model.SetTrainingMode(true);
    model.Train(input, target);

    float lastLoss = model.GetLastLoss();
    // Observed: lastLoss=0, IsNaN=False, IsInfinity=False, IsZero=True
    // Expected: NaN (CCE's log(negative_logit + eps) should produce NaN that propagates)
    Assert.False(lastLoss == 0f, "Silent zeroing of NaN loss");
}
```

Output:
```
Identity+CCE on fused-Adam: fusedSteps=1, lastLoss=0, IsNaN=False, IsInfinity=False, IsZero=True
```

## Bisection — Five Controlled Experiments

The diagnostic suite isolated the bug along five dimensions. All tests in `FlashAttentionFusedCompiledTrainingIssue1346Tests.cs` on the diagnostic branch.

| # | Setup | Expected | Actual | Status |
|---|---|---|---|---|
| 1 | Tensors-level direct: positive logits → CCE chain → `CompileTraining + Step()` | finite positive | finite positive | PASS |
| 2 | Same as #1 + `ConfigureOptimizer(Adam)` | finite positive | finite positive | PASS |
| 3 | Tensors-level 2-param model-like (matmul+add+CCE) with **negative** logits | NaN | NaN (propagated) | PASS |
| 4 | Same as #3 with **positive** logits | finite positive | finite positive | PASS |
| 5 | **AiDotNet** path (Dense+LayerNorm+SeqSlice+Dense IdentityActivation + CCE + fused Adam) | NaN (logits go negative) | **literal 0** | FAIL |

**Pattern**: Tensors-level direct usage of `CompileTraining + ConfigureOptimizer + Step` correctly propagates NaN. The AiDotNet wiring — which uses the same Tensors APIs through `CompiledTapeTrainingStep.TryStepWithFusedOptimizer` → `CompiledModelCache.GetOrCompileTraining` → `LazyTensorScope.CompileTraining` → `Step()` — fails specifically when the model has multiple trainable parameters and the loss chain produces NaN.

## What Was Ruled Out

- **Pool-padded `GetDataArray()` on the loss tensor**: I initially suspected the same pattern as the gradient-buffer bug fixed for issue #350 — the `TensorNegate` specialized forward at `CompiledTrainingPlan.cs:2030` uses `((Tensor<float>)(object)o).GetDataArray()` which returns a COPY for ArrayPool-padded tensors (logical `Length=1` on a 16-slot bucket). Writes go to the copy, reads via `lossOutput[0]` → `GetFlat` go to the live backing. The fix would mirror the gradient-buffer pattern: use `GetLiveBackingArrayAllowingPaddingOrNull() ?? GetDataArray()`. **However**: experiment #3 above traced the same negate chain via Tensors-direct and correctly returned NaN. So pool padding alone is not the cause.

- **`explicitLoss.LazySource = null` at `CompiledTrainingPlan.cs:1442` disconnecting the forward action**: doesn't apply — the lazy node still runs and writes to its `OutputBuffer` (which is the same tensor as `explicitLoss`). The disconnect only affects future lazy-graph traversals from that tensor.

- **NaN clamping in fused optimizer kernels**: `grep -rn "IsNaN" src/AiDotNet.Tensors/Engines/Compilation/CompiledTrainingPlan.cs` returns one hit at line 277 (gradient clipping max-norm), not in the loss-readout path.

- **AiDotNet-side wrapper sanitization**: `CompiledTapeTrainingStep.TryStepWithFusedOptimizer:493` reads `lossOutput[0]` directly with no NaN check. `NeuralNetworkBase.TryTrainWithFusedOptimizer:6189` assigns `LastLoss = lossValue` with no NaN check. So if Tensors hands back NaN, AiDotNet propagates NaN. The observed `lastLoss=0` means Tensors handed back 0.

## Most Likely Causes (Hypotheses for Investigation)

The diagnostic narrowed the bug to: **Tensors plan-build or plan-replay produces wrong loss readout when (a) the trace lambda includes many lazy ops in front of a CCE-style chain whose intermediate produces NaN, AND (b) `ConfigureOptimizer(Adam)` has appended a fused update closure**. Candidate causes ranked by likelihood:

1. **Forward-action pruning (`BackwardPruningPass` / `OperationReorderingPass` / etc.)** sees the loss output as unreachable from any parameter gradient consumer (the loss is only used to seed `dL/dL = 1.0`, not as a real backward input), so its forward action is dropped. The `lossGradSeed` is then re-seeded each Step with 1.0 (the array of constant ones) but the `lossOutput` tensor's storage was never written by the negate's forward — it still holds the zero/uninitialized data the pool gave it.

2. **Parameter-count-dependent fusion** in `TryFuseForwardBackward` rewrites the negate node into a fused op whose output buffer doesn't match the `explicitLoss` tensor anymore (so the negate writes to a different tensor, and `_lossOutput[0]` reads stale-zero).

3. **The loss tensor's pool-rented backing IS written by the negate's specialized forward (proven by exp #3 propagating NaN), but the AiDotNet wiring causes `lossOutput` to be re-aliased to a different tensor by some lazy-realization side effect during model layer compilation** (e.g. an in-place op on a tensor that shares storage via view-aliasing).

## How to Verify the Fix

After landing the Tensors fix:

1. Bump AiDotNet's `Directory.Packages.props` `AiDotNet.Tensors` version
2. Run `dotnet test tests/AiDotNet.Tests/AiDotNetTests.csproj --framework net10.0 -c Release --filter "FullyQualifiedName~DenseIdentity_CCE_OnFusedAdam_NaNOrZero_PinsLastLossBug"`
3. Expected: lastLoss is **NaN** (not 0). The test asserts `!isSilentlyZero` so NaN passes.

The full diagnostic test file (5 of which pin this bug from different angles) is at the AiDotNet branch `fix/1346-flashattention-compile-backward`. Cherry-pick or re-write a Tensors-level equivalent into AiDotNet.Tensors tests to enable green-after-fix gating.

## Cross-Reference

- AiDotNet#1346 — original "FlashAttentionLayer degenerate output + 3.76x slowdown" issue (engine-fix-side closed, this is the consumer-side gap that remained)
- AiDotNet.Tensors#362 — engine-side FlashAttention GraphMode branch fix (MERGED, in 0.81.3)
- AiDotNet.Tensors#367 — follow-up GraphMode audit for 5 more ops (MERGED)
- AiDotNet#1331 — fused-compiled training plan dropping gradients (FIXED, related pattern)
- AiDotNet.Tensors#350 — the original gradient-buffer pool-padding bug that established the `GetLiveBackingArrayAllowingPaddingOrNull` pattern

## Severity

**HIGH** — affects every consumer using `CategoricalCrossEntropyLoss` on raw logits with the float+Adam fused training path. The silent-failure mode (loss=0 displayed as "converging") makes this nearly impossible to diagnose without exactly the kind of multi-arm bisection above. Likely affects more than CCE — any loss whose chain can produce NaN/Inf under realistic init conditions is at risk.
